corral/util/log.pony

```diff
@@ -1,87 +1,7 @@
-type LogLevel is
-  ( LvlNone
-  | LvlErrr
-  | LvlWarn
-  | LvlInfo
-  | LvlFine
-  )
-
-primitive LvlNone
-  fun apply(): U32 => 0
-  fun string(): String => "-"
-
-primitive LvlErrr
-  fun apply(): U32 => 1
-  fun string(): String => "ERRR"
-
-primitive LvlWarn
-  fun apply(): U32 => 2
-  fun string(): String => "WARN"
-
-primitive LvlInfo
-  fun apply(): U32 => 3
-  fun string(): String => "INFO"
-
-primitive LvlFine
-  fun apply(): U32 => 4
-  fun string(): String => "FINE"
-
-primitive Level
-  fun apply(lvl: U64): LogLevel =>
-    match lvl
-    | 0 => LvlNone
-    | 1 => LvlErrr
-    | 2 => LvlWarn
-    | 3 => LvlInfo
-    | 4 => LvlFine
-    else
-      LvlFine
-    end
-
-class val Log
-  """
-  A wrapped output stream for use in logging. It supports logging at four
-  levels: err, warn, info and fine. Plus setting None to turn off output.
-  Log level is checked behind the call for convenience at the tradeoff of
-  performance.
-  The formatter can also be selected to customize the log line content and
-  format.
-  """
-  let _level: LogLevel
-  let _out: OutStream
-  let _formatter: LogFormatter
-
-  new val create(
-    level: LogLevel,
-    out: OutStream,
-    formatter: LogFormatter)
-  =>
-    _level = level
-    _out = out
-    _formatter = formatter
-
-  fun log(level: LogLevel, msg: String, loc: SourceLoc = __loc) =>
-    if (level() <= _level()) then
-      _out.print(_formatter(level, msg, loc))
-    end
-
-  fun fine(msg: String, loc: SourceLoc = __loc) =>
-    log(LvlFine, msg, loc)
-
-  fun info(msg: String, loc: SourceLoc = __loc) =>
-    log(LvlInfo, msg, loc)
-
-  fun warn(msg: String, loc: SourceLoc = __loc) =>
-    log(LvlWarn, msg, loc)
-
-  fun err(msg: String, loc: SourceLoc = __loc) =>
-    log(LvlErrr, msg, loc)
-
-interface val LogFormatter
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String
+use "logger"
 
 primitive CodeLogFormatter is LogFormatter
-  fun apply(level: LogLevel, msg: String, loc: SourceLoc): String =>
```

Comment:
I assume the log level where was so it would be polymorphic. Did you see that there's no need for being able to switch between these formatters?

Reply:
Log level was there because log.pony had its own implementation of `LogFormatter` which accepted a log level argument. The log level argument was used in `LevelLogFormatter` to prepend the log message type to the message being logged.

Reply:
Yup that makes sense now. I like how you moved the log level to being purely at the top level. 

Some package level comments for the logging subsystem for corral that details how the different parts work together should be added the PR so that folks in the future won't be asking the same questions.